Detect artists in text to disambiguate matches

okjuan opened the pull request:
Resolves #5
Resolves #8

okjuan committed:
Clean up existing integration test

okjuan committed:
Create (failing) test for goal functionality

okjuan committed:
Add tests for _find_first_two_quotes

okjuan committed:
Instantiate document parser

okjuan committed:
Search songs while parsing and match with mentioned artists

okjuan committed:
Add comments

jack-work commented on song_scrounger/document_parser.py:
```diff
@@ -1,49 +1,128 @@
+        paragraphs = text.split("\n")
+        return [p for p in paragraphs if len(p.strip(" ")) > 0]
+
+    def find_quoted_tokens(self, text):
```
Does `return re.findall("\"([^\"]*)\"", text)` work?  https://docs.python.org/3/library/re.html#re.findall

okjuan replied:
🙇 🙇‍♂️ 🙇‍♂️

jack-work commented on song_scrounger/document_parser.py:
```diff
@@ -1,49 +1,128 @@
+
+    def filter_by_mentioned_artist(self, songs, text):
+        songs_whose_artists_are_mentioned = set()
+        for song in songs:
```
Cool thing I learned when playing around with list comprehension just now: set comprehension
`return { song if self.is_mentioned(artist, text) else None for song in songs for artist in song.artist }`

No reason to do this with stateful stuff but you can have this anywhere in place of loops which just build a list (slightly more performant too I believe).

jack-work commented on song_scrounger/document_parser.py:
```diff
@@ -1,49 +1,128 @@
+    def _get_paragraphs(self, text):
+        "Returns non-empty paragraphs with one or more non-whitespace characters."
+        paragraphs = text.split("\n")
+        return [p for p in paragraphs if len(p.strip(" ")) > 0]
```
You can also use Regex for this but I can figure it out

okjuan replied:
I think this one is probably fine as-is

okjuan committed:
Create small intermediate function

IsaacSahle commented on song_scrounger/document_parser.py:
```diff
@@ -1,49 +1,128 @@
+    def __init__(self, spotify_client):
+        self.spotify_client = spotify_client
+
+    async def find_songs(self, text):
```
nit: consider validating text to not be None

okjuan committed:
Merge results when a song is repeated

okjuan committed:
Use set comprehension instead of loop

okjuan committed:
Support token-based artist name matching

jack-work commented on song_scrounger/document_parser.py:
```diff
@@ -1,49 +1,128 @@
+            for song_name in song_names:
+                songs = await self.search_spotify(song_name)
+                songs_with_mentioned_artists = self.filter_by_mentioned_artist(songs, text)
+                if len(songs_with_mentioned_artists) > 0:
```
If you define this function
`get_songs_in_text_or_default(self, song_name, text):`
`""" gets spotify songs by song_name and returns a set of those which are in the text.  If none are found in the text, it returns all songs.  """`
`songs = await self.search_spotify(song_name)`
`return filter_by_mentioned_artist(self, songs, text) or songs`

You could write this function in one line
`{ song_name: song.spotify_uri for paragraph in self._get_paragraphs(text) for song_name in self.find_quoted_tokens(paragraph) for song in await self.gets_songs_in_text_or_default(song_name, text) }`

okjuan replied:
I actually pulled out this if block into a separate function. I'll push up the changes shortly.

I personally don't like using complicated comprehensions, I think they start giving diminishing returns.

jack-work commented on song_scrounger/document_parser.py:
```diff
@@ -1,49 +1,128 @@
+                songs.append(Song(
+                    track.name,
+                    track.uri,
+                    [artist.name for artist in track.artists]
```
Something is not right here...  It looks like you're just adding |track.artists| number of Song objects to the songs list, but you're making it O(|track.artists|^2) by doing that list comprehension in the constructor.

At the very least you could store the list in a variable outside the artist in track.artists loop so you could re-use it (or you could write a function which returns a list of identical Song objects from a track based on the number of artists on the track object), but I'm curious as to why you're adding duplicates here anyway.

okjuan replied:
Yeah the logic is busted here, I've fixed it locally and pushed up the changes.

okjuan committed:
Use regex for finding tokens

okjuan committed:
Use 'song' instead of 'track' where possible

okjuan committed:
Move logic for cleaning song names

okjuan committed:
Break down existing SongScrounger class
- Move playlist creation entirely to spotify client
- Replace deprecated logic with new document_parser functionality

okjuan committed:
Move logic into SongScrounger class
